test/Libraries/Microsoft.Extensions.AI.Evaluation.Tests/ConversationTrimmingTests.cs

```diff
@@ -0,0 +1,198 @@
+
+        const int TokenBudget = 25;
+        int retainedTokens = message3.Text.Length + message2.Text.Length;
+
+        (IEnumerable<ChatMessage> trimmedConversation, int remainingTokenBudget) =
+            conversation.Trim(tokenizer, TokenBudget);
+
+        List<ChatMessage> trimmedConversationList = trimmedConversation.ToList();
+        Assert.Equal(2, trimmedConversationList.Count);
+        Assert.Equal(message3, trimmedConversationList[0]);
+        Assert.Equal(message2, trimmedConversationList[1]);
+        Assert.Equal(TokenBudget - retainedTokens, remainingTokenBudget);
+    }
+
+    [Fact]
+    public void EveryMessageInConversationExceedsTokenBudget()
+    {
+        Tokenizer tokenizer = new AsciiTokenizer();
+        var message1 = "Message 1".ToUserMessage();
+        var message2 = "Message 22".ToAssistantMessage();
+        IEnumerable<ChatMessage> conversation = [message1, message2];
+
+        const int TokenBudget = 5;
+
+        (IEnumerable<ChatMessage> trimmedConversation, int remainingTokenBudget) =
+            conversation.Trim(tokenizer, TokenBudget);
+
+        Assert.Empty(trimmedConversation);
+        Assert.Equal(TokenBudget, remainingTokenBudget);
+    }
+
+    [Fact]
+    public void LastMessageInConversationFitsTokenBudgetExactly()
+    {
+        Tokenizer tokenizer = new AsciiTokenizer();
+        var message1 = "Message 1".ToUserMessage();
+        var message2 = "Message 22".ToAssistantMessage();
+        var message3 = "Message 333".ToUserMessage();
+        IEnumerable<ChatMessage> conversation = [message1, message2, message3];
+
+        const int TokenBudget = 11;
+
+        (IEnumerable<ChatMessage> trimmedConversation, int remainingTokenBudget) =
+            conversation.Trim(tokenizer, TokenBudget);
+
+        List<ChatMessage> trimmedConversationList = trimmedConversation.ToList();
+        Assert.Single(trimmedConversationList);
+        Assert.Equal(message3, trimmedConversationList[0]);
+        Assert.Equal(0, remainingTokenBudget);
+    }
+
+    [Fact]
+    public void SubsetOfMessagesInConversationFitTokenBudgetExactly()
+    {
+        Tokenizer tokenizer = new AsciiTokenizer();
+        var message1 = "Message 1".ToUserMessage();
+        var message2 = "Message 22".ToAssistantMessage();
+        var message3 = "Message 333".ToUserMessage();
+        IEnumerable<ChatMessage> conversation = [message1, message2, message3];
+
+        const int TokenBudget = 21;
+
+        (IEnumerable<ChatMessage> trimmedConversation, int remainingTokenBudget) =
+            conversation.Trim(tokenizer, TokenBudget);
+
+        List<ChatMessage> trimmedConversationList = trimmedConversation.ToList();
+        Assert.Equal(2, trimmedConversationList.Count);
+        Assert.Equal(message3, trimmedConversationList[0]);
+        Assert.Equal(message2, trimmedConversationList[1]);
+        Assert.Equal(0, remainingTokenBudget);
+    }
+}
```

Comment:
These tests are great. Can we add a test that checks that the order of the trimmed messages is the same as the input order?

Reply:
This no longer applies since we decided to remove the helper and the above tests to avoid the dependency on Microsoft.ML.Tokenizers entirely.